### DISCUSSION TITLE
The frequency deviation of sx1262 is large

### DISCUSSION THREAD

**Discussion by weekroom** [jgromes/RadioLib#1146] [Q&A]
At present, I am using esp32 +sx1262 for lora test, one is responsible for sending and the other is responsible for receiving. Although the data can be received normally in the work, the frequency deviation of printing return is too large, I don't know how to reduce this deviation. The distance between the two is less than ten centimeters
// LoRa config
#define LORA_SEND_TEST 0
#define LORA_SEND_INTERVAL 5000

#if LORA_LONGRANGE_CONFIG == 1
#define LORA_BW 62.5
#define LORA_SF 12
#define LORA_CR 5
#else
#define LORA_BW 500.0
#define LORA_SF 7
#define LORA_CR 5
#endif
#define LORA_FREQ 915.0
#define LORA_SYNC_WORD 0x20
#define LORA_TX_POWER 22
#define LORA_PREAMBLE_LEN 20
int state = radio.begin(LORA_FREQ, LORA_BW, LORA_SF, LORA_CR, LORA_SYNC_WORD, LORA_TX_POWER, LORA_PREAMBLE_LEN, 0.0, true);
    if (state == RADIOLIB_ERR_NONE)
    {
        Serial.printf("success! %.1f %3.2f %d %d %d %d %d\r\n",
                         LORA_FREQ,
                         LORA_BW,
                         LORA_SF,
                         LORA_CR,
                         LORA_SYNC_WORD,
                         LORA_TX_POWER,
                         LORA_PREAMBLE_LEN);
    }
    else
    {
        Serial.printf("LoRa Init failed, code: %d\r\n", state);
        while (1)
        {
            delay(100);
        }
    }
static void lora_recv_test()
{
    if (lora_rx_flag)
    {
        lora_rx_flag = false;
        lora_count++;

        String lora_rx_str;
        int state = radio.readData(lora_rx_str);

        if (state == RADIOLIB_ERR_NONE)
        {
            // packet was successfully received
            Serial.println(F("[SX1262] Received packet!"));

            // print data of the packet
            Serial.print(F("[SX1262] Data:\t\t"));
            Serial.println(lora_rx_str);

            // print RSSI (Received Signal Strength Indicator)
            Serial.print(F("[SX1262] RSSI:\t\t"));
            Serial.print(radio.getRSSI());
            Serial.println(F(" dBm"));

            // print SNR (Signal-to-Noise Ratio)
            Serial.print(F("[SX1262] SNR:\t\t"));
            Serial.print(radio.getSNR());
            Serial.println(F(" dB"));

            // print frequency error
            Serial.print(F("[SX1262] Frequency error:\t"));
            Serial.print(radio.getFrequencyError());
            Serial.println(F(" Hz"));
           
            if (sd_card_init_flag)
            {
                File file = SD.open("/lora_recv.txt", FILE_APPEND);
                if (file)
                {
                    file.printf("[LoRa]\r\n%.*s\r\n", lora_rx_str.length(), lora_rx_str.c_str());
                    file.printf("rssi: %.1f dBm, snr: %.2f dB\r\n\r\n", radio.getRSSI(), radio.getSNR());
                }
                file.flush();
                file.close();
            }
        }
        else if (state == RADIOLIB_ERR_CRC_MISMATCH)
        {
            // packet was received, but is malformed
            Serial.println(F("CRC error!"));
            canvas.printf("[LoRa]\r\nCRC error!\r\n\r\n");
        }
        else
        {
            // some other error occurred
            Serial.print(F("failed, code "));
            Serial.println(state);
           
        }

        Serial.println();
    }
}

static void lora_send_test()
{
    String lora_tx_str = "Hello World! Count: " + String(lora_count++);
    int state = radio.transmit(lora_tx_str);

    if (state == RADIOLIB_ERR_NONE)
    {
        Serial.println("Send success");
        
    }
    else
    {
        Serial.printf("Send fail, code: %d\r\n", state);

    }
}
The following is receiving the printed data：
[SX1262] Received packet!
[SX1262] Data:          Hello World!Count: 670
[SX1262] RSSI:          -17.00 dBm
[SX1262] SNR:           5.75 dB
[SX1262] Frequency error:       -1918.25 Hz

[SX1262] Received packet!
[SX1262] Data:          Hello World!  Count: 673
[SX1262] RSSI:          -17.00 dBm
[SX1262] SNR:           5.25 dB
[SX1262] Frequency error:       -1917.76 Hz

[SX1262] Received packet!
[SX1262] Data:          Hello World! Count: 675
[SX1262] RSSI:          -17.00 dBm
[SX1262] SNR:           5.50 dB
[SX1262] Frequency error:       -1918.73 Hz

**Comment by HeadBoffin** [comment]
> the frequency deviation of printing return is too large

What are you comparing this to? It's pretty consistent and 0.00021% different from the set value which is pretty amazing for the price of a LoRa module, when I was a lad we'd use hot gravel to keep the radio sets at stable temperature.

> The distance between the two is less than ten centimeters

LoRa stands for Long Range, you have them 10cm apart and transmitting at full power so I'd expect issues on the receiver input circuitry as you are shouting directly in to the ear of the receiver. That may have some bearing on the frequency error.

This isn't an issue with RadioLib - it's setting up the radio with the parameters you want and then reporting what it sees. Let us know what you think.

**Reply by weekroom** [reply]
I also tried to test at a certain distance, but the actual test result was the same as that of the close distance, and the deviation of frequency was always within the range of -2000~-4000hz. What I want to know is that this should not be normal data, and the allowable range of frequency deviation is only a few hundred hz

**Reply by weekroom** [reply]
My current idea is how to reduce its frequency deviation, if the frequency deviation is too large, it is easier to receive data when testing long-distance communication

**Reply by jgromes** [reply]
This is a perfectly acceptable frequency offset - LoRa can handle up to 20% of the signal bandwidth without any issues. In your case, that's 100 kHz.

**Reply by weekroom** [reply]
Well, thank you for your answers
